File: tomviz/pipeline/PythonNodeBackend.cxx
/* This source file is part of the Tomviz project, https://tomviz.org/.
   It is released under the 3-Clause BSD License, see "LICENSE". */

#include "PythonNodeBackend.h"

#include "InputPort.h"
#include "Node.h"
#include "OutputPort.h"
#include "PythonNodeUtils.h"
#include "PythonNodeWrapper.h"
#include "data/VolumeData.h"

#pragma push_macro("slots")
#undef slots
#include <pybind11/embed.h>
#include <pybind11/pybind11.h>

#include "pybind11/PybindVTKTypeCaster.h"
#pragma pop_macro("slots")

#include <vtkImageData.h>
#include <vtkMolecule.h>
#include <vtkNew.h>
#include <vtkPythonUtil.h>
#include <vtkSmartPointer.h>
#include <vtkTable.h>

#include <QJsonDocument>

PYBIND11_VTK_TYPECASTER(vtkImageData)

namespace py = pybind11;

namespace tomviz {
namespace pipeline {

namespace {

// Names of the user-facing Python base classes in tomviz.nodes. The
// backend imports these to hand to PythonNodeUtils::findNodeClass when
// dispatching to the user's class.
constexpr const char* kSourceBaseAttr = "SourceNode";
constexpr const char* kTransformBaseAttr = "TransformNode";

/// Convert a PortData payload into a Python object suitable for the
/// inputs dict. ImageData/Volume/TiltSeries get wrapped in a fresh
/// internal_dataset.Dataset (around a deep copy of the input vtkImageData, so
/// in-place mutations by the user's transform don't corrupt the
/// upstream port's payload). Tables/molecules pass through as raw
/// vtkObjects via vtkPythonUtil.
py::object portDataToPython(const PortData& data, py::object datasetCls)
{
  if (!data.isValid()) {
    return py::none();
  }
  PortType type = data.type();
  if (isVolumeType(type)) {
    auto vol = data.value<VolumeDataPtr>();
    if (!vol || !vol->isValid()) {
      return py::none();
    }
    vtkNew<vtkImageData> copy;
    copy->DeepCopy(vol->imageData());
    return datasetCls(py::cast(static_cast<vtkImageData*>(copy.Get()),
                               py::return_value_policy::reference));
  }
  if (type == PortType::Table) {
    auto sp = data.value<vtkSmartPointer<vtkTable>>();
    if (!sp) {
      return py::none();
    }
    return py::reinterpret_steal<py::object>(
      vtkPythonUtil::GetObjectFromPointer(sp.GetPointer()));
  }
  if (type == PortType::Molecule) {
    auto sp = data.value<vtkSmartPointer<vtkMolecule>>();
    if (!sp) {
      return py::none();
    }
    return py::reinterpret_steal<py::object>(
      vtkPythonUtil::GetObjectFromPointer(sp.GetPointer()));
  }
  return py::none();
}

} // namespace

PythonNodeBackend::PythonNodeBackend() = default;

void PythonNodeBackend::setJSONDescription(const QString& json)
{
  m_jsonDescription = json;
  parseDescription();
}

QString PythonNodeBackend::jsonDescription() const
{
  return m_jsonDescription;
}

void PythonNodeBackend::setScript(const QString& script)
{
  m_script = script;
}

QString PythonNodeBackend::scriptSource() const
{
  return m_script;
}

QString PythonNodeBackend::operatorName() const { return m_operatorName; }
QString PythonNodeBackend::defaultLabel() const { return m_defaultLabel; }
QString PythonNodeBackend::descriptionText() const { return m_description; }
QString PythonNodeBackend::helpText() const { return m_help; }
QString PythonNodeBackend::customWidgetID() const { return m_customWidgetID; }
bool PythonNodeBackend::supportsCancel() const { return m_supportsCancel; }
bool PythonNodeBackend::supportsComplete() const { return m_supportsComplete; }
bool PythonNodeBackend::isTransformShape() const { return !m_inputs.isEmpty(); }

QString PythonNodeBackend::externalPythonEnvPath() const
{
  return m_externalPythonEnvPath;
}

void PythonNodeBackend::setParameter(const QString& name,
                                     const QVariant& value)
{
  m_parameters[name] = value;
}

QVariant PythonNodeBackend::parameter(const QString& name) const
{
  return m_parameters.value(name);
}

QMap<QString, QVariant> PythonNodeBackend::parameters() const
{
  return m_parameters;
}

QStringList PythonNodeBackend::inputNames() const
{
  QStringList names;
  for (const auto& p : m_inputs) {
    names.append(p.name);
  }
  return names;
}

QStringList PythonNodeBackend::outputNames() const
{
  QStringList names;
  for (const auto& p : m_outputs) {
    names.append(p.name);
  }
  return names;
}

QString PythonNodeBackend::primaryOutputName() const
{
  return m_outputs.isEmpty() ? QString() : m_outputs.first().name;
}

void PythonNodeBackend::parseDescription()
{
  m_operatorName.clear();
  m_defaultLabel.clear();
  m_description.clear();
  m_help.clear();
  m_customWidgetID.clear();
  m_supportsCancel = false;
  m_supportsComplete = false;
  m_externalPythonEnvPath.clear();
  m_inputs.clear();
  m_outputs.clear();
  m_parameters.clear();
  m_parameterTypes.clear();
  m_enumOptions.clear();

  QJsonDocument doc = QJsonDocument::fromJson(m_jsonDescription.toUtf8());
  if (!doc.isObject()) {
    return;
  }
  QJsonObject obj = doc.object();

  m_operatorName = obj.value(QStringLiteral("name")).toString();
  m_defaultLabel = obj.value(QStringLiteral("label")).toString();
  m_description = obj.value(QStringLiteral("description")).toString();
  m_help = obj.value(QStringLiteral("help")).toString();
  m_customWidgetID = obj.value(QStringLiteral("widget")).toString();
  m_supportsCancel =
    obj.value(QStringLiteral("supportsCancel")).toBool(false);
  m_supportsComplete =
    obj.value(QStringLiteral("supportsComplete")).toBool(false);
  m_externalPythonEnvPath =
    obj.value(QStringLiteral("tomviz_pipeline_env")).toString();

  // inputs / outputs are arrays of {name, type[, persistent]}. Missing
  // section → empty list (per the agreed schema-v2 convention).
  for (const auto& v : obj.value(QStringLiteral("inputs")).toArray()) {
    QJsonObject entry = v.toObject();
    PortSpec spec;
    spec.name = entry.value(QStringLiteral("name")).toString();
    spec.type =
      portTypeFromString(entry.value(QStringLiteral("type")).toString());
    if (!spec.name.isEmpty() && spec.type != PortType::None) {
      m_inputs.append(spec);
    }
  }
  for (const auto& v : obj.value(QStringLiteral("outputs")).toArray()) {
    QJsonObject entry = v.toObject();
    PortSpec spec;
    spec.name = entry.value(QStringLiteral("name")).toString();
    spec.type =
      portTypeFromString(entry.value(QStringLiteral("type")).toString());
    spec.persistent =
      entry.value(QStringLiteral("persistent")).toBool(false);
    if (!spec.name.isEmpty() && spec.type != PortType::None) {
      m_outputs.append(spec);
    }
  }

  // parameters: same shape and coercion semantics as schema-v1, so the
  // shared PythonNodeUtils helpers handle int/double/enum without
  // Qt6's QJsonValue→QVariant<double> collapse breaking type-sensitive
  // operators.
  for (const auto& v : obj.value(QStringLiteral("parameters")).toArray()) {
    QJsonObject param = v.toObject();
    QString name = param.value(QStringLiteral("name")).toString();
    QString type = param.value(QStringLiteral("type")).toString();
    QJsonValue defaultVal = param.value(QStringLiteral("default"));
    if (name.isEmpty()) {
      continue;
    }
    m_parameterTypes[name] = type;

    if (type == QLatin1String("enumeration")) {
      QJsonArray options = param.value(QStringLiteral("options")).toArray();
      m_enumOptions[name] = options;
      QVariant resolved =
        PythonNodeUtils::resolveEnumValue(defaultVal, options);
      if (resolved.isValid()) {
        m_parameters[name] = resolved;
        continue;
      }
    }

    QVariant value =
      PythonNodeUtils::coerceJsonByDeclaredType(defaultVal, type);
    if (!value.isValid()) {
      // Complex / unknown declared type with no explicit default: let
      // the user's Python default win. With an explicit default fall
      // back to QJsonValue's own conversion (preserves string / list).
      if (defaultVal.isUndefined() || defaultVal.isNull()) {
        continue;
      }
      value = defaultVal.toVariant();
    }
    m_parameters[name] = value;
  }
}

void PythonNodeBackend::applyDescription(AddInputFn addInput,
                                         AddOutputFn addOutput)
{
  if (addInput) {
    for (const auto& spec : m_inputs) {
      addInput(spec.name, spec.type);
    }
  }
  if (addOutput) {
    for (const auto& spec : m_outputs) {
      OutputPort* port = addOutput(spec.name, spec.type);
      if (port) {
        // Schema-v2 convention: omitted "persistent" → false. Set
        // explicitly so the v2 default holds even though the C++
        // OutputPort default is true.
        port->setPersistent(spec.persistent);
      }
    }
  }
}

QJsonObject PythonNodeBackend::serializeInto(QJsonObject base) const
{
  base[QStringLiteral("description")] = m_jsonDescription;
  base[QStringLiteral("script")] = m_script;
  if (!m_parameters.isEmpty()) {
    base[QStringLiteral("arguments")] =
      QJsonObject::fromVariantMap(m_parameters);
  }
  return base;
}

void PythonNodeBackend::applySerializedFields(const QJsonObject& json,
                                              AddInputFn addInput,
                                              AddOutputFn addOutput)
{
  if (json.contains(QStringLiteral("description"))) {
    setJSONDescription(json.value(QStringLiteral("description")).toString());
  }
  if (json.contains(QStringLiteral("script"))) {
    setScript(json.value(QStringLiteral("script")).toString());
  }
  applyDescription(std::move(addInput), std::move(addOutput));

  auto args = json.value(QStringLiteral("arguments")).toObject();
  for (auto it = args.constBegin(); it != args.constEnd(); ++it) {
    const QString& key = it.key();
    const QString declType = m_parameterTypes.value(key);
    QVariant qv;
    if (declType == QLatin1String("enumeration")) {
      qv = PythonNodeUtils::resolveEnumValue(
        it.value(), m_enumOptions.value(key));
    }
    if (!qv.isValid()) {
      qv = PythonNodeUtils::coerceJsonByDeclaredType(it.value(), declType);
    }
    if (!qv.isValid()) {
      qv = it.value().toVariant();
    }
    m_parameters[key] = qv;
  }
}

QMap<QString, PortData> PythonNodeBackend::runTransform(
  Node* host, const QMap<QString, PortData>& inputs)
{
  return runImpl(host, inputs, /*isSource=*/false);
}

QMap<QString, PortData> PythonNodeBackend::runSource(Node* host)
{
  return runImpl(host, {}, /*isSource=*/true);
}

QMap<QString, PortData> PythonNodeBackend::runImpl(
  Node* host, const QMap<QString, PortData>& inputs, bool isSource)
{
  QMap<QString, PortData> result;
  if (!host) {
    return result;
  }

  // Ensure the embedded interpreter is alive. Never finalize — C
  // extension modules like numpy can't be re-loaded after
  // finalize_interpreter().
  if (!Py_IsInitialized()) {
    py::initialize_interpreter();
  }

  try {
    py::gil_scoped_acquire gil;

    // Some user scripts may reach into tomviz.utils.* without
    // importing it (legacy carry-over).
    py::module_::import("tomviz.utils");

    py::module_ datasetMod = py::module_::import("tomviz.internal_dataset");
    py::object datasetCls = datasetMod.attr("Dataset");

    py::module_ nodesMod = py::module_::import("tomviz.nodes");
    py::object baseClass =
      nodesMod.attr(isSource ? kSourceBaseAttr : kTransformBaseAttr);

    py::object scriptModule =
      PythonNodeUtils::loadScriptAsModule(m_operatorName, m_script);

    py::object userClass =
      PythonNodeUtils::findNodeClass(scriptModule, baseClass);
    if (userClass.is_none()) {
      qWarning("PythonNodeBackend: no %s subclass found in script",
               isSource ? "SourceNode" : "TransformNode");
      return result;
    }

    // Instantiate following the operator-class pattern: __new__ → set
    // _operator_wrapper → __init__. Lets the user's __init__ access
    // self.progress / self.canceled / self.completed.
    py::object wrapper =
      createNodeWrapper(host, primaryOutputName());
    py::object instance = userClass.attr("__new__")(userClass);
    instance.attr("_operator_wrapper") = wrapper;
    userClass.attr("__init__")(instance);

    // Build kwargs from current parameters.
    py::dict kwargs;
    for (auto it = m_parameters.constBegin();
         it != m_parameters.constEnd(); ++it) {
      kwargs[py::str(it.key().toStdString())] =
        PythonNodeUtils::qvariantToPython(it.value());
    }

    py::object pyResult;
    if (isSource) {
      pyResult = instance.attr("produce")(**kwargs);
    } else {
      // Convert each input PortData to the right Python representation.
      // Volume-shaped inputs get a deep-copied internal_dataset.Dataset wrapper
      // so user mutation doesn't leak upstream.
      py::dict inputsDict;
      for (auto it = inputs.constBegin(); it != inputs.constEnd(); ++it) {
        inputsDict[py::str(it.key().toStdString())] =
          portDataToPython(it.value(), datasetCls);
      }
      pyResult =
        instance.attr("transform")(inputsDict, **kwargs);
    }

    if (host->isCanceled()) {
      return result;
    }

    if (pyResult.is_none() || !py::isinstance<py::dict>(pyResult)) {
      // None is the documented return for "cancellation or error" per
      // tomviz.nodes — the user's transform/produce returned without
      // producing outputs. Any other non-dict return is treated the
      // same way. The caller's port-empty check transitions the node
      // to Failed.
      return result;
    }

    // Pull each declared output by name. Unknown / missing keys are
    // ignored (the caller will detect missing outputs via the
    // node-level check that PortData was set on every output port).
    py::dict outDict = pyResult.cast<py::dict>();
    for (const auto& spec : m_outputs) {
      OutputPort* port = host->outputPort(spec.name);
      if (!port) {
        continue;
      }
      std::string key = spec.name.toStdString();
      if (!outDict.contains(key)) {
        continue;
      }
      py::object payload = outDict[py::str(key)];
      PortData pd =
        PythonNodeUtils::pythonValueToPortData(payload, port);
      if (pd.isValid()) {
        result[spec.name] = pd;
      }
    }
  } catch (const py::error_already_set& e) {
    qWarning("PythonNodeBackend Python error: %s", e.what());
  } catch (const std::exception& e) {
    qWarning("PythonNodeBackend error: %s", e.what());
  }

  return result;
}

} // namespace pipeline
} // namespace tomviz

File: tomviz/python/tomviz/pipeline/python_node_backend.py
###############################################################################
# This source file is part of the Tomviz project, https://tomviz.org/.
# It is released under the 3-Clause BSD License, see "LICENSE".
###############################################################################
"""Shared backend for schema-v2 Python source / transform nodes in the CLI
runtime. Mirrors the C++ ``tomviz::pipeline::PythonNodeBackend``: owns the
JSON description, the script source, and the current parameter values; runs
the user's ``tomviz.nodes.SourceNode`` / ``tomviz.nodes.TransformNode``
subclass when the host's shell calls ``run_source`` / ``run_transform``.

The shells (``PythonSource`` in ``sources/python_source.py`` and
``PythonTransform`` in ``transforms/python_transform.py``) are thin —
they own one of these and forward."""

from __future__ import annotations

import copy
import importlib.util
import inspect
import json
import logging
import os
import tempfile
from typing import Any, Callable

# Some user scripts reach into tomviz.utils.* without importing it
# (legacy carry-over).
import tomviz.utils  # noqa: F401

from tomviz._internal import OperatorWrapper
from tomviz.pipeline.node import PortData


logger = logging.getLogger('tomviz.pipeline')


def _coerce_param_default(param_type: str, default):
    if isinstance(default, list):
        return list(default)
    if param_type == 'double':
        return float(default) if default is not None else 0.0
    if param_type in ('int', 'integer', 'enumeration'):
        return int(default) if default is not None else 0
    if param_type in ('bool', 'boolean'):
        return bool(default)
    if param_type in ('string', 'file', 'save_file', 'directory'):
        return str(default) if default is not None else ''
    return default


def _resolve_enum_default(param: dict):
    """For an enumeration parameter, return the value of the option at
    the index given by ``default``, or ``None`` if the description is
    malformed."""
    options = param.get('options') or []
    default = param.get('default')
    if not isinstance(default, int) or not (0 <= default < len(options)):
        return None
    opt = options[default]
    if not isinstance(opt, dict) or not opt:
        return None
    return next(iter(opt.values()))


def _find_node_class(module, base_class):
    """Locate a single subclass of ``base_class`` in ``module``. Returns
    ``None`` when no subclass is present; raises ``ValueError`` when more
    than one is found (mirrors the C++
    ``PythonNodeUtils::findNodeClass`` contract)."""
    found = None
    for _, cls in inspect.getmembers(module, inspect.isclass):
        if cls is base_class:
            continue
        if not issubclass(cls, base_class):
            continue
        if found is not None:
            raise ValueError(
                'Multiple node classes defined in module — only one '
                'node class can be defined per script.')
        found = cls
    return found


class PythonNodeBackend:
    """Owns the schema-v2 Python state and runs the user's class."""

    def __init__(self):
        self.json_description: str = ''
        self.script: str = ''
        self.operator_name: str = ''
        self.default_label: str = ''
        self.description_text: str = ''
        self.help_text: str = ''
        self.custom_widget_id: str = ''
        self.supports_cancel: bool = False
        self.supports_complete: bool = False
        self.external_python_env_path: str = ''
        # List of (name, port_type) for inputs.
        self._inputs: list[tuple[str, str]] = []
        # List of (name, port_type, persistent) for outputs.
        self._outputs: list[tuple[str, str, bool]] = []
        self.parameters: dict[str, Any] = {}
        self._parameter_types: dict[str, str] = {}
        self._enum_options: dict[str, list] = {}

    # ---- accessors ----------------------------------------------------

    def primary_output_name(self) -> str:
        return self._outputs[0][0] if self._outputs else ''

    def is_transform_shape(self) -> bool:
        return bool(self._inputs)

    # ---- description / script -----------------------------------------

    def set_json_description(self, json_str: str):
        self.json_description = json_str
        self._parse_description()

    def set_script(self, script: str):
        self.script = script

    def _parse_description(self):
        self.operator_name = ''
        self.default_label = ''
        self.description_text = ''
        self.help_text = ''
        self.custom_widget_id = ''
        self.supports_cancel = False
        self.supports_complete = False
        self.external_python_env_path = ''
        self._inputs = []
        self._outputs = []
        self.parameters = {}
        self._parameter_types = {}
        self._enum_options = {}

        if not self.json_description:
            return
        try:
            obj = json.loads(self.json_description)
        except json.JSONDecodeError:
            logger.exception('Failed to parse operator JSON description')
            return
        if not isinstance(obj, dict):
            return

        self.operator_name = obj.get('name', '')
        self.default_label = obj.get('label', '')
        self.description_text = obj.get('description', '')
        self.help_text = obj.get('help', '')
        self.custom_widget_id = obj.get('widget', '')
        self.supports_cancel = bool(obj.get('supportsCancel', False))
        self.supports_complete = bool(obj.get('supportsComplete', False))
        self.external_python_env_path = obj.get('tomviz_pipeline_env', '')

        for entry in obj.get('inputs', []) or []:
            name = entry.get('name')
            ptype = entry.get('type')
            if name and ptype:
                self._inputs.append((name, ptype))
        for entry in obj.get('outputs', []) or []:
            name = entry.get('name')
            ptype = entry.get('type')
            persistent = bool(entry.get('persistent', False))
            if name and ptype:
                self._outputs.append((name, ptype, persistent))

        for param in obj.get('parameters', []) or []:
            name = param.get('name')
            ptype = param.get('type', '')
            if not name:
                continue
            self._parameter_types[name] = ptype
            if ptype == 'enumeration':
                options = param.get('options') or []
                self._enum_options[name] = options
                resolved = _resolve_enum_default(param)
                if resolved is not None:
                    self.parameters[name] = resolved
                    continue
            if 'default' not in param:
                continue
            self.parameters[name] = _coerce_param_default(
                ptype, param['default'])

    # ---- port creation -------------------------------------------------

    def apply_description(
            self,
            add_input: Callable[[str, str], None] | None,
            add_output: Callable[[str, str, bool], None] | None):
        if add_input:
            for name, ptype in self._inputs:
                add_input(name, ptype)
        if add_output:
            for name, ptype, persistent in self._outputs:
                add_output(name, ptype, persistent)

    # ---- serialize / deserialize --------------------------------------

    def serialize_into(self, base: dict) -> dict:
        base['description'] = self.json_description
        base['script'] = self.script
        if self.parameters:
            base['arguments'] = dict(self.parameters)
        return base

    def apply_serialized_fields(
            self, data: dict,
            add_input: Callable[[str, str], None] | None,
            add_output: Callable[[str, str, bool], None] | None):
        if 'description' in data:
            self.set_json_description(data['description'])
        if 'script' in data:
            self.script = data.get('script', '')
        self.apply_description(add_input, add_output)
        for key, value in (data.get('arguments') or {}).items():
            self.parameters[key] = value

    # ---- execution ----------------------------------------------------

    def run_transform(self, host, inputs: dict[str, PortData]
                      ) -> dict[str, PortData]:
        return self._run_impl(host, inputs, is_source=False)

    def run_source(self, host) -> dict[str, PortData]:
        return self._run_impl(host, {}, is_source=True)

    def _run_impl(self, host, inputs: dict[str, PortData],
                  is_source: bool) -> dict[str, PortData]:
        from tomviz import nodes
        base_class = nodes.SourceNode if is_source else nodes.TransformNode

        module = self._load_script_module()
        if module is None:
            return {}

        try:
            user_class = _find_node_class(module, base_class)
        except ValueError:
            logger.exception(
                'Multiple %s subclasses found in script for %s',
                'SourceNode' if is_source else 'TransformNode',
                self.operator_name)
            return {}
        if user_class is None:
            logger.error(
                'No %s subclass found in script for %s',
                'SourceNode' if is_source else 'TransformNode',
                self.operator_name)
            return {}

        # Instantiate; tomviz.nodes.Node.__new__ pre-attaches a
        # tomviz.operators.Progress(self), which we then overwrite with
        # the CLI's progress object so `self.progress.value = X` lands
        # in the parent-bound progress channel directly (bypassing the
        # _operator_wrapper indirection that's unused on the CLI side).
        instance = user_class()
        progress = getattr(host, 'progress', None)
        if progress is not None:
            instance.progress = progress
            channel = (progress.control_channel()
                       if hasattr(progress, 'control_channel') else None)
            instance._operator_wrapper = OperatorWrapper(channel)
            if hasattr(progress, 'set_primary_port'):
                progress.set_primary_port(self.primary_output_name())
        else:
            instance._operator_wrapper = OperatorWrapper(None)

        kwargs = dict(self.parameters)

        try:
            if is_source:
                result = instance.produce(**kwargs)
            else:
                # Deep-copy each input payload so user mutations don't
                # leak back into upstream port state. Mirrors the C++
                # backend's portDataToPython(deep-copy) behavior.
                inputs_dict = {
                    name: copy.deepcopy(pd.payload)
                    for name, pd in inputs.items()
                }
                result = instance.transform(inputs_dict, **kwargs)
        except Exception:
            logger.exception("Operator '%s' raised", self.operator_name)
            return {}

        # None is the documented signal for "cancel or error" per
        # tomviz.nodes — the user's transform/produce returned without
        # producing outputs. Any other non-dict is treated the same way.
        if not isinstance(result, dict):
            return {}

        outputs: dict[str, PortData] = {}
        for name, ptype, _ in self._outputs:
            if name in result:
                outputs[name] = PortData(result[name], ptype)
        return outputs

    def _load_script_module(self):
        """Materialize self.script as a temp .py and import it. Returns
        the module, or ``None`` on failure."""
        fd = None
        path = None
        try:
            fd, path = tempfile.mkstemp(suffix='.py', text=True)
            os.write(fd, self.script.encode())
            os.close(fd)
            fd = None
            label = (self.operator_name
                     or self.default_label
                     or 'NodeModule')
            spec = importlib.util.spec_from_file_location(label, path)
            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)
            return module
        except Exception:
            logger.exception(
                'Failed to load script for %s', self.operator_name)
            return None
        finally:
            if fd is not None:
                os.close(fd)
            if path is not None:
                try:
                    os.unlink(path)
                except OSError:
                    pass
